hal/armv7a/zynq7000/zynq.c

```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
+			*clkact0 = *clkact1 = *divisor0 = *divisor1 = 0;
+			break;
+
+		case pctl_ctrl_gem0_rclk:
+		case pctl_ctrl_gem1_rclk:
+			id = dev - pctl_ctrl_gem0_rclk;
+			val = *(zynq_common.slcr + slcr_gem0_rclk_ctrl + id);
+			*clkact0 = val & 0x1;
+			*srcsel = (val >> 4) & 0x1;
+			*clkact1 = *divisor0 = *divisor1 = 0;
+			return 0;
+
+		case pctl_ctrl_gem0_clk:
+		case pctl_ctrl_gem1_clk:
+			id = dev - pctl_ctrl_gem0_clk;
+			val = *(zynq_common.slcr + slcr_gem0_clk_ctrl + id);
+			*clkact0 = val & 0x1;
+			*srcsel = (val >> 4) & 0x7;
+			*divisor0 = (val >> 8) & 0x3f;
+			*divisor1 = (val >> 20) & 0x3f;
+			*clkact1 = 0;
+			break;
+
+		case pctl_ctrl_smc_clk:
+			val = *(zynq_common.slcr + slcr_smc_clk_ctrl);
+			*clkact0 = val & 0x1;
+			*srcsel = (val >> 4) & 0x3;
+			*divisor0 = (val >> 8) & 0x3f;
+			*clkact1 = *divisor1 = 0;
+			break;
+
+		case pctl_ctrl_lqspi_clk:
+			val = *(zynq_common.slcr + slcr_lqspi_clk_ctrl);
+			*clkact0 = val & 0x1;
+			*srcsel = (val >> 4) & 0x3;
+			*divisor0 = (val >> 8) & 0x3f;
+			*clkact1 = *divisor1 = 0;
+			break;
+
+		case pctl_ctrl_sdio_clk:
+			val = *(zynq_common.slcr + slcr_sdio_clk_ctrl);
+			*clkact0 = val & 0x1;
+			*clkact1 = (val >> 1) & 0x1;
+			*srcsel = (val >> 4) & 0x3;
+			*divisor0 = (val >> 8) & 0x3f;
+			*divisor1 = 0;
+			break;
+
+		case pctl_ctrl_uart_clk:
+			val = *(zynq_common.slcr + slcr_uart_clk_ctrl);
+			*clkact0 = val & 0x1;
+			*clkact1 = (val >> 1) & 0x1;
+			*srcsel = (val >> 4) & 0x3;
+			*divisor0 = (val >> 8) & 0x3f;
+			*divisor1 = 0;
+			break;
+
+		case pctl_ctrl_spi_clk:
+			val = *(zynq_common.slcr + slcr_spi_clk_ctrl);
+			*clkact0 = val & 0x1;
+			*clkact1 = (val >> 1) & 0x1;
+			*srcsel = (val >> 4) & 0x3;
+			*divisor0 = (val >> 8) & 0x3f;
+			*divisor1 = 0;
+			break;
+
+		case pctl_ctrl_can_clk:
+			val = *(zynq_common.slcr + slcr_can_clk_ctrl);
+			*clkact0 = val & 0x1;
+			*clkact1 = (val >> 1) & 0x1;
+			*srcsel = (val >> 4) & 0x3;
+			*divisor0 = (val >> 8) & 0x3f;
+			*divisor1 = (val >> 20) & 0x3f;
+			break;
+
+		default:
+			return -1;
+	}
+
+	return 0;
+}
+
+
+static int _zynq_setMioClk(int mio, char ref0, char mux0, char ref1, char mux1)
+{
+	if (mio == pctl_ctrl_can_mioclk) {
+		_zynq_slcrUnlock();
+		*(zynq_common.slcr + slcr_can_mioclk_ctrl) = (*(zynq_common.slcr + slcr_can_mioclk_ctrl) & ~0x007f007f) | (mux0 & 0x3f) | ((!!ref0) << 6) |
+			((mux1 & 0x3f) << 16) | ((!!ref1) << 22);
+		_zynq_slcrLock();
+
+		return 0;
+	}
+
+	return -1;
+}
+
+
+static int _zynq_getMioClk(int mio, char *ref0, char *mux0, char *ref1, char *mux1)
+{
+	u32 val = 0;
+
+	if (mio == pctl_ctrl_can_mioclk) {
+		val = *(zynq_common.slcr + slcr_can_mioclk_ctrl);
+		*mux0 = val & 0x3f;
+		*ref0 = (val >> 6) & 0x1;
+		*mux1 = (val >> 16) & 0x3f;
+		*ref1 = (val >> 22) & 0x1;
+
+		return 0;
+	}
+
+	return -1;
+}
+
+
+static int _zynq_setMIO(int pin, char disableRcvr, char pullup, char ioType, char speed, char l0, char l1, char l2, char l3, char triEnable)
```

Comment:
10 members of the `mio` struct as arguments to the function?

Reply:
Yeah, I know that it looks ugly. We use this approach to omit adding some extra structure definitions, it is internal function so it doesn't bother us so much.